utils/bitfield/src/lib.rs

```diff
@@ -1,462 +1,247 @@
-            _ => unreachable!(),
-        }
-    }
+    /// Returns an iterator over the ranges of set bits that make up the bit field. The
+    /// ranges are in ascending order, are non-empty, and don't overlap.
+    pub fn ranges(&self) -> impl RangeIterator + '_ {
+        let ranges = |set: &AHashSet<usize>| {
+            let mut vec: Vec<_> = set.iter().copied().collect();
+            vec.sort_unstable();
+            ranges_from_bits(vec)
+        };
 
-    /// Merges to bitfields together (equivalent of bitwise OR `|` operator)
-    pub fn merge(mut self, other: &Self) -> Result<Self> {
-        self.merge_assign(other)?;
-        Ok(self)
+        self.bitvec
+            .ranges()
+            .union(ranges(&self.set))
+            .difference(ranges(&self.unset))
     }
 
-    /// Merges to bitfields into `self` (equivalent of bitwise OR `|` operator)
-    pub fn merge_assign(&mut self, other: &Self) -> Result<()> {
-        let a = self.as_mut_flushed()?;
-        match other {
-            BitField::Encoded { bv, set, unset } => {
-                let v = decode_and_apply_cache(bv, set, unset)?;
-                bit_or(a, v.into_iter())
-            }
-            BitField::Decoded(bv) => bit_or(a, bv.iter().copied()),
-        }
-
-        Ok(())
+    /// Returns `true` if the bit field is empty.
+    pub fn is_empty(&self) -> bool {
+        self.set.is_empty()
+            && self
+                .bitvec
+                .ranges()
+                .flatten()
+                .all(|bit| self.unset.contains(&bit))
     }
 
-    /// Intersection of two bitfields (equivalent of bit AND `&`)
-    pub fn intersect(mut self, other: &Self) -> Result<Self> {
-        self.intersect_assign(other)?;
-        Ok(self)
-    }
+    /// Returns a slice of the bit field with the start index of set bits
+    /// and number of bits to include in the slice. Returns an error if the
+    /// bit field contains fewer than `start + len` set bits.
+    pub fn slice(&self, start: usize, len: usize) -> Result<Self> {
+        let slice = BitField::from_ranges(self.ranges().skip_bits(start).take_bits(len));
 
-    /// Intersection of two bitfields and assigns to self (equivalent of bit AND `&`)
-    pub fn intersect_assign(&mut self, other: &Self) -> Result<()> {
-        match other {
-            BitField::Encoded { bv, set, unset } => {
-                *self.as_mut_flushed()? &= decode_and_apply_cache(bv, set, unset)?
-            }
-            BitField::Decoded(bv) => *self.as_mut_flushed()? &= bv.iter().copied(),
+        if slice.len() >= len {
+            Ok(slice)
+        } else {
+            Err("Not enough bits")
         }
-        Ok(())
     }
 
-    /// Subtract other bitfield from self (equivalent of `a & !b`)
-    pub fn subtract(mut self, other: &Self) -> Result<Self> {
-        self.subtract_assign(other)?;
-        Ok(self)
+    /// Returns the number of set bits in the bit field.
+    pub fn len(&self) -> usize {
```

Comment:
Yeah I feel you, I have struggled with this in the documentation. If you look at a bit field as a sequence of 1s and 0s then `len` here is obviously incorrect and `count_ones` would make more sense. But if you look at it as a collection of indices that represent where the 1s are (like a more efficient and ordered `HashSet<usize>`), then `len` as the number of 1s makes sense again.

I think we should probably go with the "ordered collection of indices" definition because then it makes sense that:
- `first` returns an index and not the value of the first bit,
- `iter` iterates over the indices of the 1s and not over a bunch of `bool`s,
- `ranges` only cares about 1s,
- etc

Also `union`/`intersection`/`difference` treat it as a collection of 1s.

The `len`/`count` distinction is so subtle that I don't really care either way, to me `count` doesn't convey that it only cares about the set bits more than `len` does? And common Rust wisdom seems to dictate that `count` consumes the thing (like `Iterator::count`) while `len` doesn't, so that's what I based it on.

Whatever we go for, docs couldn't definitely improve in this regard.

Reply:
yeah, that's all reasonable. I'm fine with this change, so maybe see if anyone else has a preference as I have nothing else I care to add to my previous comment

Reply:
I agree with @timvermeulen here. Thinking about this as an "ordered collection of indices" makes sense for `len` insofar as it is consistent with `first`, `iter`, etc... Since the other publicly exposed functions dont really make reference to the underlying RLE encoded bits, there shouldn't be confusion there.